### ISSUE TITLE
Clarification on `num_valid_positions` in Section 4.3.5 (multinomial_loss)

### ISSUE THREAD

**Issue by ChengYe97** [instadeepai/nucleotide-transformer#125]
Hi NT team,

Thanks for your impressive work on **NTv3**!

In **Section 4.3.5**, the pseudocode for `multinomial_loss` uses a variable `num_valid_positions` (**line 17**), but it appears to be **undefined** in the text.

```python
1
2 def poisson_loss(predictions: Array, targets: Array) -> float:
3     return predictions - targets * log(predictions + epsilon)
4
5 def multinomial_loss(
6     logits: Array,
7     targets: Array,
8     shape_loss_coefficient: float = 0.2,
9 ) -> tuple[float, float, float]:
10     # Scale (Poisson) loss on total coverage
11     sum_pred = sum(logits, axis=seq_dim)
12     sum_true = sum(targets, axis=seq_dim)
13     scale_loss = mean(poisson_loss(sum_pred, sum_true))
14
15     # Shape (multinomial) loss on distribution
16     p_pred = logits / sum(logits, axis=seq_dim, keepdims=True)
17     shape_loss = -sum(targets * log(p_pred + epsilon)) / num_valid_positions
18
19     total_loss = shape_loss + shape_loss_coefficient * scale_loss
20
21     return total_loss
22
```

Could you please clarify its definition? For example, does it refer to the number non-zero values in array`targets`?

Thanks for your help!

**Comment by BDEvan5** [comment]
Hello ChengYe97

Thanks for finding the error. Here is an updated version, indicating that `num_valid_positions` (renamed to `num_elements`) is the number of positions where the loss is calculated.

```python
    def multinomial_loss(
        logits: Array,
        targets: Array,
        shape_loss_coefficient: float = 0.2,
    ) -> tuple[float, float, float]:
        # Scale (Poisson) loss on total coverage
        sum_pred = sum(logits, axis=seq_dim)
        sum_true = sum(targets, axis=seq_dim)
        scale_loss = mean(poisson_loss(sum_pred, sum_true))
    
        # Shape (multinomial) loss on distribution
        p_pred = logits / sum(logits, axis=seq_dim, keepdims=True)
        num_elements = prod(logits.shape) + epsilon
        shape_loss = -sum(targets * log(p_pred + epsilon)) / num_elements
    
        total_loss = shape_loss + shape_loss_coefficient * scale_loss

        return total_loss
```

Best regards
Benjamin

**Comment by ChengYe97** [comment]
> Hello ChengYe97
> 
> Thanks for finding the error. Here is an updated version, indicating that `num_valid_positions` (renamed to `num_elements`) is the number of positions where the loss is calculated.
> 
>     def multinomial_loss(
>         logits: Array,
>         targets: Array,
>         shape_loss_coefficient: float = 0.2,
>     ) -> tuple[float, float, float]:
>         # Scale (Poisson) loss on total coverage
>         sum_pred = sum(logits, axis=seq_dim)
>         sum_true = sum(targets, axis=seq_dim)
>         scale_loss = mean(poisson_loss(sum_pred, sum_true))
>     
>         # Shape (multinomial) loss on distribution
>         p_pred = logits / sum(logits, axis=seq_dim, keepdims=True)
>         num_elements = prod(logits.shape) + epsilon
>         shape_loss = -sum(targets * log(p_pred + epsilon)) / num_elements
>     
>         total_loss = shape_loss + shape_loss_coefficient * scale_loss
> 
>         return total_loss
> Best regards Benjamin

Thank you for the update and your response. I’ve got it now. Best regards!

**Comment by ChengYe97** [comment]
@BDEvan5  Additionally, I noticed a potential typo in **Section 4.3.5 (Post-training Objective and Loss Function)** regarding the variable definitions. While the text states that $x$ are the labels and $y$ are the predictions, **Equations (2) and (3)** appear to swap these roles. 

<img width="804" height="398" alt="Image" src="https://github.com/user-attachments/assets/663e7e22-d637-4bcb-ac6e-933bbc1a9447" />

I am not sure if this has been addressed in your latest version, but it seems that **$x$ should represent the predictions and $y$ the labels** for the formulas to be consistent. Thank you again for your outstanding work, and I hope this helps in your review of the paper.

Best regards

**Comment by BDEvan5** [comment]
You are correct, that is a typo and the variable descriptions in the main text should be swopped. 

Thanks for finding that and commenting.
